### PR TITLE
Sign up

### PR DESCRIPTION
**Pull request by alex6296**
-added a sign up view
-added validation on inputs
-added a way to call post user on our server

### CODE REVIEW

File: src/views/SignUp/SignUp.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    this.handleSubmitFields = this.handleSubmitFields.bind(this);
+    this.VerifyString = this.VerifyString.bind(this);
+    this.VerifyEmail = this.VerifyEmail.bind(this);
+    this.verifyPassword = this.verifyPassword.bind(this);
+    this.comparePassword = this.comparePassword.bind(this);
+  }
+
+  handleSubmitFields(event) {
+
+    if(this.state.includes(null)){
```
Comment (kvartborg) [inline src/views/SignUp/SignUp.js]: this doesn't make sense, the state will never include `null`?

File: src/views/SignUp/SignUp.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+    this.setState({
+      [name]: target.value
+    });
+  }
+
+  render(){
+    return (
+      <Layout>
+        <h1>Sign up</h1>
```
Comment (grammeaway) [inline src/views/SignUp/SignUp.js]: Uppercase U in Up would be sexier 🦅

File: src/views/SignUp/SignUp.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    }).then(res => res.json())
+    .then(response => console.log('Success:', JSON.stringify(response)))
+    .catch(error => console.error('Error:', error));
+    event.preventDefault();
+  }
+
+  VerifyString({ target }) {
+    const {name} = target;
+    if(name.length<=0){
+      console.log('is empty')
```
Comment (kvartborg) [inline src/views/SignUp/SignUp.js]: Remove all `console.log`, they should not be committed as their only purpose is for debugging.

File: src/views/SignUp/SignUp.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+      return;
+    }
+    console.log(name+' exsists')
+
+    this.setState({
+      [name]: target.value
+    });
+  }
+
+  VerifyEmail({ target }) {
```
Comment (kvartborg) [inline src/views/SignUp/SignUp.js]: Methods in js shall be camelCased.

File: src/views/SignUp/SignUp.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            </label>
+
+            <label>
+              Password
+              <input type="password" name="password" min="8" onChange={this.verifyPassword} required/>
+            </label>
+
+            <label>
+              Verify Password
+              <input type="password" name="verifyPassword"  min="8" onChange={this.comparePassword} required/>
```
Comment (grammeaway) [inline src/views/SignUp/SignUp.js]: Just tested the verification of password, it accepts different passwords within the two fields

File: src/views/SignUp/SignUp.js (hunk 4 above)
Comment (kvartborg) [reply src/views/SignUp/SignUp.js]: This is not a valid email validation, we need to use a regex to ensure all emails are accepted.

File: src/views/SignUp/SignUp.js [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    this.setState({
+      [name]: target.value
+    });
+  }
+
+  render(){
+    return (
+      <Layout>
+        <h1>Sign up</h1>
+          <form onSubmit={this.handleSubmitFields}>
```
Comment (kvartborg) [inline src/views/SignUp/SignUp.js]: The `form` should be moved one tab back.

File: src/views/SignUp/SignUp.js (hunk 2 above)
Comment (alex6296) [reply src/views/SignUp/SignUp.js]: resolve # 4a967e1

File: src/views/Main.js [hunk 7]
```diff
@@ -6,7 +6,8 @@ export default () => (
   <Layout>
     <section>
       <h1>Main View</h1>
-      <Link to='/second'>Second view</Link>
+      <p> <Link to='/second'>Second view</Link></p>
+      <p> <Link to='/sign-up'>Sing up</Link></p>
```
Comment (kvartborg) [inline src/views/Main.js]:
```suggestion
      <p><Link to='/sign-up'>Sing up</Link></p>
```

File: src/views/Main.js [hunk 8]
```diff
@@ -6,7 +6,8 @@ export default () => (
   <Layout>
     <section>
       <h1>Main View</h1>
-      <Link to='/second'>Second view</Link>
+      <p> <Link to='/second'>Second view</Link></p>
```
Comment (kvartborg) [inline src/views/Main.js]:
```suggestion
      <p><Link to='/second'>Second view</Link></p>
```

File: src/views/SignUp/SignUp.js (hunk 2 above)
Comment (alex6296) [reply src/views/SignUp/SignUp.js]: 4a967e1

File: src/views/SignUp/SignUp.js (hunk 4 above)
Comment (kvartborg) [reply src/views/SignUp/SignUp.js]: The regex below should validate emails. Add this to a variable and run the input against the regex.
```js
/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)\|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])\|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
```

File: src/views/SignUp/SignUp.js (hunk 5 above)
Comment (alex6296) [reply src/views/SignUp/SignUp.js]: 63fae75

File: src/views/SignUp/SignUp.js [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+    fetch(url, {
+      method: 'POST',
+      body: JSON.stringify(this.state),
+      headers:{
+        'Content-Type': 'application/json'
+      }
+    }).then(res => res.json())
+    .then(response => console.log('Success:', JSON.stringify(response)))
+    .catch(error => console.error('Error:', error));
```
[the remark this answers is not in the document]
Comment (alex6296) [reply src/views/SignUp/SignUp.js]: e3c8709

File: src/views/SignUp/SignUp.js (hunk 3 above)
Comment (alex6296) [reply src/views/SignUp/SignUp.js]: 7ed41c2
1 changed file

File: src/Routes.js [hunk 10]
```diff
@@ -2,10 +2,12 @@ import React from 'react';
 import { Route } from "react-router-dom";
 import Main from './views/Main';
 import Second from './views/Second';
+import SignUp from './views/SignUp'
 import Create from './views/Task/Create';
 
 export default () => [
   <Route path='/' exact component={Main} />,
   <Route path='/second' component={Second} />,
```
Comment (grammeaway) [inline src/Routes.js]: There's a merge conflict here, be  sure to resolve it, and commit it 😄

File: src/views/SignUp/SignUp.js [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    this.addToState(name,target.value);
+  }
+
+  verifyPassword({ target }){
+    if(target.value.length<8) {
+      return;
+    }
+  }
+
+  comparePassword({ target }) {
```
Comment (grammeaway) [inline src/views/SignUp/SignUp.js]: I am still able to make a submission with two different passwords 😿

File: src/views/SignUp/SignUp.js [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  render(){
+    return (
+      <Layout>
+        <h1>Sign Up</h1>
+        <form onSubmit={this.handleSubmitFields}>
+          <label>
+            First Name
+            <input type="text" name="firstname" onChange={this.verifyString} required/>
```
Comment (kvartborg) [inline src/views/SignUp/SignUp.js]: Missing value binding.
```suggestion
            <input type="text" name="firstname" onChange={this.verifyString} value={this.state.data.firstName} required/>
```

File: src/views/SignUp/SignUp.js [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+   * @param {event}
+   */
+  submitHandler= event =>{
+    event.preventDefault();
+
+    if(!this.validator()){
+      //invalid input
+      return false;
+    }
+    const response = serverApi.post(this.state);
```
Comment (kvartborg) [inline src/views/SignUp/SignUp.js]: Unused assign.

File: src/views/SignUp/SignUp.js [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+    //["label", "Type", Validation method]
+    ["First name","text", v => v.length > 0],
+    ["Last name","text", v => v.length > 0],
+    ["Email","email", v => (/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/).test(v)], // eslint-disable-line],
+    ["Username","text", v => v.length > 0],
+    ["Password","password", v => v.length >= 8],
+    ["Verify password","password", v => v === this.state.password && v.length>=8]
+  ]
+ /**
+   * event listener for <input/>'s
```
[the remark this answers is not in the document]
Comment (alex6296) [reply src/views/SignUp/SignUp.js]: is the return type {html} acceptable or should it be something else?

File: src/api/worker.js [hunk 15]
```diff
@@ -6,7 +6,11 @@ export default {
    * @return {Promise}
    */
   async get() {
-    const res = await request.get('workers')
-    return await res.json()
+    const res = await request.get('workers');
+    return await res.json();
+  },
+  async post (body, options){
+    const res = await request.post('https://localhost:5001/api/workers/',body);
```
[the remark this answers is not in the document]
Comment (kvartborg) [reply src/api/worker.js]: The method should be named create.

File: src/views/SignUp/SignUp.js [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+   * adds changes to this.state
+   * @param  {event}
```
Comment (kvartborg) [inline src/views/SignUp/SignUp.js]: Its the type which should be described within `{}`
```suggestion
   * @param  {Object} event
```

File: src/views/SignUp/SignUp.js [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+    if(!this.validator()){
+      //invalid input
+      return false;
+    }
+    const response = serverApi.post(this.state);
+  }
+
+  /**
+   * Checks if all input fields meets their specific requirements
+   * @return {boolean}
```
Comment (kvartborg) [inline src/views/SignUp/SignUp.js]:
```suggestion
   * @return {Boolean} 
```

File: src/views/SignUp/SignUp.js [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+   */
+  changeHandler = event => {
+    this.setState({
+      [event.target.name] : event.target.value
+    })
+  }
+/**
+   * event listener for <Submit/>
+   * validates all inputfields before sending a post request to the server
+   * @param {event}
```
Comment (kvartborg) [inline src/views/SignUp/SignUp.js]:
```suggestion
   * @param {Object} event
```

File: src/views/SignUp/SignUp.js [hunk 19]
```diff
@@ -0,0 +1,105 @@
+import React,{Component} from "react";
+import Layout from '../../components/Layout';
+import camelcase from 'camelcase';
+import serverApi from '../../api/worker';
```
Comment (kvartborg) [inline src/views/SignUp/SignUp.js]: Always name the imports as the filename.
```suggestion
import worker from '../../api/worker';
```

File: src/views/SignUp/SignUp.js [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+   * @param {Object} event
+   */
+  submitHandler= event =>{
+    event.preventDefault();
+
+    if(!this.validator()){
+      //invalid input
+      return false;
+    }
+    serverApi.create(this.state);
```
Comment (kvartborg) [inline src/views/SignUp/SignUp.js]:
```suggestion
    worker.create(this.state);
```

File: src/views/SignUp/SignUp.js [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+  ]
+
+ /**
+   * event listener for input
+   * adds changes to this.state
+<<<<<<< HEAD
+   * @param  {object} event
+=======
+   * @param  {Object} event
+>>>>>>> 7c90e501ae0a04d9e3f9b0a8ae07801a693803a6
```
Comment (kvartborg) [inline src/views/SignUp/SignUp.js]: Fix conflicts

File: src/views/SignUp/SignUp.js [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+/**
+   * event listener for Submit
+   * validates all inputfields before sending a post request to the server
+<<<<<<< HEAD
+   * @param {object} event
+=======
+   * @param {Object} event
+>>>>>>> 7c90e501ae0a04d9e3f9b0a8ae07801a693803a6
```
Comment (kvartborg) [inline src/views/SignUp/SignUp.js]: Fix conflicts

File: src/views/SignUp/SignUp.js [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+          value={this.state[name]}
+          onChange={this.changeHandler}
+          className={validator(this.state[name]) ? "valid": ""}
+          required
+        />
+      </label>
+    )
+  }
+
+    /**
```
Comment (kvartborg) [inline src/views/SignUp/SignUp.js]:
```suggestion
   /**
```

**Pull request merged**